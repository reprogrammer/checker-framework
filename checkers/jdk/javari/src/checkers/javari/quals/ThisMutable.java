package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation that indicates that the type of a field is the same as the
 * type of the reference via which the field was accessed.
 * <p>
 *
 * For example, consider a Bicycle class that that contains a Wheel object:
 * <pre>
 * class Bicycle {
 *   Wheel frontWheel;
 *   ...
 * }
 * </pre>
 *
 * In a mutable bicycle, it is reasonable to mutate the wheel.  In a
 * read-only bicycle, the wheel should not be mutated -- it should be
 * read-only just like the bicyle that contains it.  "This-mutability"
 * means thath the field has the same mutability as the <tt>this</tt>
 * reference.
 * <pre>
 * @Readonly Bicycle b1;
 * @Mutable  Bicycle b2;
 * ... b1.frontWheel ...    // type is: @Readonly Wheel
 * ... b2.frontWheel ...    // type is: @Mutable Wheel
 * </pre>
 * 
 * <tt>@ThisMutable</tt> is the default on fields, and does not make sense
 * to write elsewhere.  Therefore, <tt>@ThisMutable</tt> should never
 * appear in a program.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
