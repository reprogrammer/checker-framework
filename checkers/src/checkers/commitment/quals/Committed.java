package checkers.commitment.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.DefaultQualifierInHierarchy;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@TypeQualifier
@SubtypeOf(Unclassified.class)
@DefaultQualifierInHierarchy
public @interface Committed {
}
