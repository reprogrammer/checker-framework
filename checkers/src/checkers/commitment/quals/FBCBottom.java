package checkers.commitment.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * {@link FBCBottom} marks the bottom of the Freedom Before Commitment type
 * hierarchy. It cannot be used in client code and is only used internally.
 *
 * @author Stefan Heule
 */
@Documented
@TypeQualifier
@SubtypeOf({ Free.class, Committed.class })
@Retention(RetentionPolicy.RUNTIME)
// empty target prevents programmers from writing this in a program
@Target({})
public @interface FBCBottom {
}
