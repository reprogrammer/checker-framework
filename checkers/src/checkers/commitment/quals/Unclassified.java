package checkers.commitment.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@TypeQualifier
@SubtypeOf({})
public @interface Unclassified {
}
