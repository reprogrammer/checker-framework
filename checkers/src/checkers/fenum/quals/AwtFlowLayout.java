package checkers.fenum.quals;

import java.lang.annotation.*;

import checkers.quals.*;

/**
 * Line alignments in a flow layout (see
 * {@link java.awt.FlowLayout} for more details).
 * @see java.awt.FlowLayout#LEFT
 * @see java.awt.FlowLayout#CENTER
 * @see java.awt.FlowLayout#RIGHT
 * @see java.awt.FlowLayout#LEADING
 * @see java.awt.FlowLayout#TRAILING
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE_PARAMETER, ElementType.TYPE_USE})
@TypeQualifier
@SubtypeOf(FenumTop.class)
public @interface AwtFlowLayout {}
