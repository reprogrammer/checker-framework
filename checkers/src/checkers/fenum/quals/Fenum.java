package checkers.fenum.quals;

import java.lang.annotation.*;

import checkers.quals.*;

/**
 * A generic fake enumeration qualifier that is parameterized by a name.
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE_PARAMETER, ElementType.TYPE_USE})
@TypeQualifier
@SubtypeOf(FenumTop.class)
public @interface Fenum {
    String value();
}
