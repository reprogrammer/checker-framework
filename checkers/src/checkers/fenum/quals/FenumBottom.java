package checkers.fenum.quals;

import java.lang.annotation.*;

import checkers.quals.ImplicitFor;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

import com.sun.source.tree.Tree;

/**
 * The bottom of the type hierarchy.
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target( {} )
@TypeQualifier
@SubtypeOf( { Fenum.class, FenumUnqualified.class } )
@ImplicitFor(trees={Tree.Kind.NULL_LITERAL})
public @interface FenumBottom {}
