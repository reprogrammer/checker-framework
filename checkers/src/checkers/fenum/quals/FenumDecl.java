package checkers.fenum.quals;

import java.lang.annotation.*;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;


/**
 * 
 * @author wmdietl
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD})
@TypeQualifier
@SubtypeOf( { FenumTop.class } )
public @interface FenumDecl {
    String value() default "";
}