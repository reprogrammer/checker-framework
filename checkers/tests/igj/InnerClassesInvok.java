import checkers.igj.quals.*;

@I
class A {
    void isAMutable(@Mutable A this) { }
    void isAImmutable(@Immutable A this) { }
    void isAReadOnly(@ReadOnly A this) { }

    void mutableMethodForA(@Mutable A this) {
        @I
        class B {
            void isBMutable(@Mutable B this) { }
            void isBImmutable(@Immutable B this) { }
            void isBReadOnly(@ReadOnly B this) { }

            void testImmutableForB(@Immutable B this) {
                isAMutable();
                isAImmutable(); // error
                isAReadOnly();
                A.this.isAMutable();
                A.this.isAImmutable(); // error
                A.this.isAReadOnly();

                isBMutable();   // error
                isBImmutable();
                isBReadOnly();

                B.this.isBMutable();   // error
                B.this.isBImmutable();
                B.this.isBReadOnly();
            }

            void testMutableForB(@Mutable B this) {
                isAMutable();
                isAImmutable(); // error
                isAReadOnly();
                A.this.isAMutable();
                A.this.isAImmutable(); // error
                A.this.isAReadOnly();

                isBMutable();
                isBImmutable(); // error
                isBReadOnly();

                B.this.isBMutable();
                B.this.isBImmutable();  // error
                B.this.isBReadOnly();
            }

        }
    }
}

