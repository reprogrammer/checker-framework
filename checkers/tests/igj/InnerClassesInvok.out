:18: error: (method.invocation.invalid)
:21: error: (method.invocation.invalid)
:24: error: (method.invocation.invalid)
:28: error: (method.invocation.invalid)
:35: error: (method.invocation.invalid)
:38: error: (method.invocation.invalid)
:42: error: (method.invocation.invalid)
:46: error: (method.invocation.invalid)
